File: Java/ZCLI.java
// ZCLI.java - Z.AI CLI Java Version

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import java.util.zip.*;

public class ZCLI {
    // ANSI color codes
    private static final String CYAN = "\033[36m";
    private static final String GREEN = "\033[32m";
    private static final String YELLOW = "\033[33m";
    private static final String RED = "\033[31m";
    private static final String GRAY = "\033[90m";
    private static final String RESET = "\033[0m";
    
    private static boolean inCodeBlock = false;
    private static StringBuilder fullText = new StringBuilder();
    
    private static class Model {
        String id;
        String display;
        Model(String id, String display) {
            this.id = id;
            this.display = display;
        }
    }
    
    private static class File {
        String path;
        String content;
        File(String path, String content) {
            this.path = path;
            this.content = content;
        }
    }
    
    private static Map<String, Model> availableModels = new HashMap<>();
    static {
        availableModels.put("1", new Model("zai-org/GLM-5", "GLM 5"));
        availableModels.put("2", new Model("zai-org/GLM-5.1", "Z GLM 5.1"));
    }
    
    private static String question(String prompt) throws IOException {
        System.out.print(prompt);
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        return reader.readLine();
    }
    
    private static String post(String host, String path, String data) throws IOException {
        URL url = new URL("https://" + host + path);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        
        try (OutputStream os = conn.getOutputStream()) {
            os.write(data.getBytes());
            os.flush();
        }
        
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) {
                response.append(line);
            }
            return response.toString();
        }
    }
    
    private static List<File> streamWithFilter(String host, String path, String data) throws IOException {
        URL url = new URL("https://" + host + path);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        conn.setChunkedStreamingMode(0);
        
        try (OutputStream os = conn.getOutputStream()) {
            os.write(data.getBytes());
            os.flush();
        }
        
        List<File> allFiles = new ArrayList<>();
        
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.trim().isEmpty()) continue;
                
                // Parse JSON line
                if (line.contains("\"choices\"")) {
                    String text = extractContentFromJson(line);
                    if (text != null && !text.isEmpty()) {
                        fullText.append(text);
                        
                        // Character by character processing
                        for (int i = 0; i < text.length(); i++) {
                            if (i + 2 < text.length() && text.substring(i, i+3).equals("```")) {
                                inCodeBlock = !inCodeBlock;
                                i += 2;
                            } else if (!inCodeBlock) {
                                System.out.print(text.charAt(i));
                            }
                        }
                    }
                }
            }
        }
        
        // Extract files using regex
        Pattern pattern = Pattern.compile("```\\w*\\{path=([^}]+)\\}\\n([\\s\\S]*?)```");
        Matcher matcher = pattern.matcher(fullText.toString());
        
        while (matcher.find()) {
            String filePath = matcher.group(1);
            String content = matcher.group(2).replaceAll("```$", "").trim();
            allFiles.add(new File(filePath, content));
            System.out.println(CYAN + " Make: " + filePath + RESET);
        }
        
        return allFiles;
    }
    
    private static String extractContentFromJson(String json) {
        try {
            // Simple extraction without full JSON parser
            Pattern pattern = Pattern.compile("\"content\":\"([^\"]*)\"");
            Matcher matcher = pattern.matcher(json);
            if (matcher.find()) {
                return matcher.group(1);
            }
        } catch (Exception e) {}
        return null;
    }
    
    private static void writeFiles(List<File> files, Path outDir) throws IOException {
        for (File file : files) {
            Path fullPath = outDir.resolve(file.path);
            Files.createDirectories(fullPath.getParent());
            Files.write(fullPath, file.content.getBytes());
            System.out.println(GREEN + "✓ " + file.path + RESET);
        }
    }
    
    private static Model selectModel() throws IOException {
        System.out.println("\n" + CYAN + " Select an AI model:" + RESET);
        System.out.println("  1. GLM 5");
        System.out.println("  2. Z GLM 5.1");
        String choice = question(YELLOW + " Select (1-2, default: 1): " + RESET);
        choice = choice.trim().isEmpty() ? "1" : choice;
        return availableModels.getOrDefault(choice, availableModels.get("1"));
    }
    
    public static void main(String[] args) throws IOException {
        System.out.println("\n" + CYAN + "╔════════════════════════════════════╗" + RESET);
        System.out.println(CYAN + "║                Z.AI CLI                       ║" + RESET);
        System.out.println(CYAN + "╚════════════════════════════════════╝" + RESET + "\n");
        
        Model selectedModel = selectModel();
        System.out.println(GREEN + "✓ Model: " + selectedModel.display + RESET + "\n");
        
        String prompt = question(YELLOW + "Prompt Input: " + RESET);
        if (prompt.trim().isEmpty()) {
            System.out.println(RED + "✗ Prompt cannot be empty!" + RESET);
            return;
        }
        
        String defaultFolder = prompt.toLowerCase().replaceAll("\\s+", "-");
        if (defaultFolder.length() > 30) defaultFolder = defaultFolder.substring(0, 30);
        String folderName = question(YELLOW + " Destination folder name (default: " + defaultFolder + "): " + RESET);
        String finalFolder = folderName.trim().isEmpty() ? defaultFolder : folderName;
        
        System.out.println("\n" + CYAN + " Select output mode:" + RESET);
        System.out.println("  1. Save to folder (without zip)");
        System.out.println("  2. Save to folder + zip");
        String mode = question(YELLOW + " Choose (1/2, default: 1): " + RESET);
        
        boolean isZip = mode.trim().equals("2");
        Path outDir = Paths.get(finalFolder);
        Path zipPath = Paths.get(finalFolder + ".zip");
        
        System.out.println("\n" + CYAN + "Generating: " + prompt + RESET + "\n");
        
        try {
            // Create chat
            String chatData = String.format(
                "{\"prompt\":\"%s\",\"model\":\"%s\",\"quality\":\"low\"}",
                prompt, selectedModel.id
            );
            String response = post("llamacoder.together.ai", "/api/create-chat", chatData);
            
            // Extract lastMessageId (simple parsing)
            String lastMessageId = extractValueFromJson(response, "lastMessageId");
            
            System.out.println(GREEN + "✓ Chat created" + RESET);
            System.out.println(YELLOW + " AI is responding..." + RESET + "\n");
            System.out.println(GRAY + "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" + RESET + "\n");
            
            String streamData = String.format(
                "{\"messageId\":%s,\"model\":\"%s\"}",
                lastMessageId, selectedModel.id
            );
            
            List<File> files = streamWithFilter("llamacoder.together.ai", 
                "/api/get-next-completion-stream-promise", streamData);
            
            System.out.println("\n" + GRAY + "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" + RESET + "\n");
            
            if (files.isEmpty()) {
                System.out.println(RED + "✗ No files are generated" + RESET);
                return;
            }
            
            System.out.println(GREEN + " Total " + files.size() + " file will be created" + RESET + "\n");
            
            String confirm = question(YELLOW + "Continue writing file? (y/n, default: y): " + RESET);
            if (confirm.trim().toLowerCase().equals("n")) {
                System.out.println(RED + "✗ Canceled" + RESET);
                return;
            }
            
            if (Files.exists(outDir)) {
                System.out.println(YELLOW + "Folder " + finalFolder + " already exists, it will be overwritten..." + RESET);
                deleteDirectory(outDir.toFile());
            }
            
            Files.createDirectories(outDir);
            
            System.out.println("\n" + CYAN + " Write files to disk..." + RESET + "\n");
            writeFiles(files, outDir);
            
            System.out.println("\n" + GREEN + " Succeed! " + files.size() + " the file has been saved" + RESET);
            
            if (isZip) {
                System.out.println(CYAN + " Create zip..." + RESET);
                try {
                    zipDirectory(outDir.toFile(), zipPath.toFile());
                    System.out.println(GREEN + "✓ Zip created successfully: " + zipPath + RESET);
                    
                    String deleteFolder = question(YELLOW + " Delete folder after zipping? (y/n, default: y): " + RESET);
                    if (!deleteFolder.trim().toLowerCase().equals("n")) {
                        deleteDirectory(outDir.toFile());
                        System.out.println(GREEN + "✓ Folder " + finalFolder + " deleted" + RESET);
                    } else {
                        System.out.println(GREEN + "✓ Permanent folders are stored in: " + outDir + RESET);
                    }
                } catch (Exception e) {
                    System.out.println(YELLOW + " Failed to create zip, files remain in folder: " + outDir + RESET);
                }
            } else {
                System.out.println(GREEN + " Done! The file is saved in: " + outDir + RESET);
            }
            
        } catch (Exception e) {
            System.out.println(RED + "✗ Error: " + e.getMessage() + RESET);
        }
    }
    
    private static String extractValueFromJson(String json, String key) {
        Pattern pattern = Pattern.compile("\"" + key + "\":(\\d+)");
        Matcher matcher = pattern.matcher(json);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return "0";
    }
    
    private static void deleteDirectory(File dir) {
        if (dir.isDirectory()) {
            File[] children = dir.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteDirectory(child);
                }
            }
        }
        dir.delete();
    }
    
    private static void zipDirectory(File sourceDir, File zipFile) throws IOException {
        try (ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(zipFile))) {
            Path sourcePath = sourceDir.toPath();
            Files.walk(sourcePath)
                .filter(path -> !Files.isDirectory(path))
                .forEach(path -> {
                    ZipEntry zipEntry = new ZipEntry(sourcePath.relativize(path).toString());
                    try {
                        zos.putNextEntry(zipEntry);
                        Files.copy(path, zos);
                        zos.closeEntry();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                });
        }
    }
}

File: Rust/zcli.rs
// zcli.rs - Z.AI CLI Rust Version

use reqwest;
use serde_json::{json, Value};
use std::io::{self, Write, BufRead};
use std::fs;
use std::path::PathBuf;
use regex::Regex;
use colored::*;

#[derive(Debug)]
struct Model {
    id: String,
    display: String,
}

struct File {
    path: String,
    content: String,
}

fn question(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

async fn post(host: &str, path: &str, data: Value) -> Result<Value, Box<dyn std::error::Error>> {
    let url = format!("https://{}{}", host, path);
    let client = reqwest::Client::new();
    let response = client.post(&url)
        .json(&data)
        .send()
        .await?;
    let json = response.json::<Value>().await?;
    Ok(json)
}

async fn stream_with_filter(host: &str, path: &str, data: Value) -> Result<Vec<File>, Box<dyn std::error::Error>> {
    let url = format!("https://{}{}", host, path);
    let client = reqwest::Client::new();
    let mut response = client.post(&url)
        .json(&data)
        .send()
        .await?;
    
    let mut full_text = String::new();
    let mut in_code_block = false;
    let mut all_files = Vec::new();
    
    while let Some(chunk) = response.chunk().await? {
        let chunk_str = String::from_utf8_lossy(&chunk);
        let lines: Vec<&str> = chunk_str.split('\n').collect();
        
        for line in lines {
            if let Ok(json) = serde_json::from_str::<Value>(line) {
                if let Some(text) = json["choices"][0]["delta"]["content"].as_str() {
                    full_text.push_str(text);
                    
                    // Character by character processing
                    let chars: Vec<char> = text.chars().collect();
                    let mut i = 0;
                    while i < chars.len() {
                        if i + 2 < chars.len() && chars[i] == '`' && chars[i+1] == '`' && chars[i+2] == '`' {
                            in_code_block = !in_code_block;
                            i += 3;
                        } else if !in_code_block {
                            print!("{}", chars[i]);
                            io::stdout().flush().unwrap();
                            i += 1;
                        } else {
                            i += 1;
                        }
                    }
                }
            }
        }
    }
    
    // Extract files using regex
    let re = Regex::new(r"```\w*\{path=([^}]+)\}\n([\s\S]*?)```")?;
    for cap in re.captures_iter(&full_text) {
        let file_path = cap[1].to_string();
        let content = cap[2].to_string().trim_end_matches('`').to_string();
        println!("{} Make: {}", "Make:".cyan(), file_path.cyan());
        all_files.push(File { path: file_path, content });
    }
    
    Ok(all_files)
}

fn write_files(files: &[File], out_dir: &PathBuf) {
    for file in files {
        let full_path = out_dir.join(&file.path);
        if let Some(parent) = full_path.parent() {
            fs::create_dir_all(parent).unwrap();
        }
        fs::write(&full_path, &file.content).unwrap();
        println!("{} ✓ {}", "✓".green(), file.path.green());
    }
}

fn select_model() -> Model {
    println!("\n{} Select an AI model:", "".cyan());
    println!("  1. GLM 5");
    println!("  2. Z GLM 5.1");
    
    let choice = question(&format!("{} Select (1-2, default: 1): {}", "".yellow(), "".reset()));
    let choice = choice.trim();
    if choice == "2" {
        Model { id: "zai-org/GLM-5.1".to_string(), display: "Z GLM 5.1".to_string() }
    } else {
        Model { id: "zai-org/GLM-5".to_string(), display: "GLM 5".to_string() }
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("\n{}╔════════════════════════════════════╗", "".cyan());
    println!("{}║                Z.AI CLI                       ║", "".cyan());
    println!("{}╚════════════════════════════════════╝{}\n", "".cyan(), "".reset());
    
    let selected_model = select_model();
    println!("{}✓ Model: {}{}\n", "✓".green(), selected_model.display.green(), "".reset());
    
    let prompt = question(&format!("{}Prompt Input: {}", "".yellow(), "".reset()));
    if prompt.trim().is_empty() {
        println!("{}✗ Prompt cannot be empty!{}", "✗".red(), "".reset());
        return Ok(());
    }
    
    let default_folder = prompt.to_lowercase().replace(" ", "-");
    let default_folder = if default_folder.len() > 30 { &default_folder[..30] } else { &default_folder };
    let folder_name = question(&format!("{} Destination folder name (default: {}): {}", "".yellow(), default_folder, "".reset()));
    let final_folder = if folder_name.trim().is_empty() { default_folder.to_string() } else { folder_name };
    
    println!("\n{} Select output mode:", "".cyan());
    println!("  1. Save to folder (without zip)");
    println!("  2. Save to folder + zip");
    let mode = question(&format!("{} Choose (1/2, default: 1): {}", "".yellow(), "".reset()));
    
    let is_zip = mode.trim() == "2";
    let out_dir = PathBuf::from(&final_folder);
    let zip_path = format!("{}.zip", final_folder);
    
    println!("\n{}Generating: {}{}\n", "".cyan(), prompt, "".reset());
    
    // Create chat
    let chat_data = json!({
        "prompt": prompt,
        "model": selected_model.id,
        "quality": "low"
    });
    
    let response = post("llamacoder.together.ai", "/api/create-chat", chat_data).await?;
    let last_message_id = response["lastMessageId"].as_i64().unwrap();
    
    println!("{}✓ Chat created{}", "✓".green(), "".reset());
    println!("{} AI is responding...{}\n", "".yellow(), "".reset());
    println!("{}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{}\n", "".bright_black(), "".reset());
    
    let stream_data = json!({
        "messageId": last_message_id,
        "model": selected_model.id
    });
    
    let files = stream_with_filter("llamacoder.together.ai", "/api/get-next-completion-stream-promise", stream_data).await?;
    
    println!("\n{}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{}\n", "".bright_black(), "".reset());
    
    if files.is_empty() {
        println!("{}✗ No files are generated{}", "✗".red(), "".reset());
        return Ok(());
    }
    
    println!("{} Total {} file will be created{}\n", "".green(), files.len().to_string().green(), "".reset());
    
    let confirm = question(&format!("{}Continue writing file? (y/n, default: y): {}", "".yellow(), "".reset()));
    if confirm.trim().to_lowercase() == "n" {
        println!("{}✗ Canceled{}", "✗".red(), "".reset());
        return Ok(());
    }
    
    if out_dir.exists() {
        println!("{}Folder {} already exists, it will be overwritten...{}", "".yellow(), final_folder.yellow(), "".reset());
        fs::remove_dir_all(&out_dir)?;
    }
    
    fs::create_dir_all(&out_dir)?;
    
    println!("\n{} Write files to disk...{}\n", "".cyan(), "".reset());
    write_files(&files, &out_dir);
    
    println!("\n{} Succeed! {} the file has been saved{}", "".green(), files.len().to_string().green(), "".reset());
    
    if is_zip {
        println!("{} Create zip...{}", "".cyan(), "".reset());
        // Zip functionality would need external crate
        println!("{} Done! The file is saved in: {}{}", "".green(), out_dir.display().to_string().green(), "".reset());
    } else {
        println!("{} Done! The file is saved in: {}{}", "".green(), out_dir.display().to_string().green(), "".reset());
    }
    
    Ok(())
}